plugin-lifecycle/src/main/java/com/mapbox/maps/plugin/lifecycle/MapboxLifecyclePluginImpl.kt

```diff
@@ -10,58 +10,57 @@ import android.content.ComponentCallbacks2.TRIM_MEMORY_RUNNING_MODERATE
 import android.content.ComponentCallbacks2.TRIM_MEMORY_UI_HIDDEN
 import android.content.res.Configuration
 import android.view.View
-import androidx.lifecycle.Lifecycle
-import androidx.lifecycle.LifecycleObserver
-import androidx.lifecycle.OnLifecycleEvent
-import androidx.lifecycle.ViewTreeLifecycleOwner
+import androidx.lifecycle.*
 import com.mapbox.maps.MapboxLifecycleObserver
-import com.mapbox.maps.logE
 import com.mapbox.maps.logW
 import com.mapbox.maps.plugin.Plugin.Companion.MAPBOX_LIFECYCLE_PLUGIN_ID
 import com.mapbox.maps.plugin.delegates.MapPluginProviderDelegate
 
 /**
  * Concrete implementation of MapboxLifecyclePlugin.
  */
-class MapboxLifecyclePluginImpl : MapboxLifecyclePlugin {
+class MapboxLifecyclePluginImpl : MapboxLifecyclePlugin, LifecycleOwner {
+  private lateinit var viewLifecycleRegistry: ViewLifecycleRegistry
+
   /**
    * Register a MapboxLifecycleObserver to observe life cycle events from LifecycleOwner
    *
    * @param mapView the instance of mapView, will get the LifecycleOwner from mapview's parent
    * @param observer the observer that listen to the life cycle events
    */
   override fun registerLifecycleObserver(mapView: View, observer: MapboxLifecycleObserver) {
-    val lifecycleOwner = ViewTreeLifecycleOwner.get(mapView)
-    if (lifecycleOwner == null) {
-      logE(
-        TAG,
+    viewLifecycleRegistry = ViewLifecycleRegistry(
+      view = mapView,
+      localLifecycleOwner = this,
+      hostingLifecycleOwner = requireNotNull(ViewTreeLifecycleOwner.get(mapView)) {
```

Comment:
This seems unnecessarily late. Aren't there cases where `getLifecycle` could be called before the view is attached to window? Couldn't we pass the `mapView` to the constructor of the plugin and setup the registry right-away? Also we don't need the view tree to find the lifecycle owner, we just need the context - https://github.com/mapbox/mapbox-navigation-android/blob/889e9f383a3777aef700df8b720e5120da6b7c1f/libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/internal/extensions/ContextEx.kt#L21.

Reply:
> This seems unnecessarily late. Aren't there cases where getLifecycle could be called before the view is attached to window? Couldn't we pass the mapView to the constructor of the plugin and setup the registry right-away? 

@LukasPaczos this is due to our plugin setup, where we don't have direct access to MapView inside the lifecycle plugin. And currently we do only call the `getLifecycle` and `registerLifecycleObserver` [when the MapView is attached to window](https://github.com/mapbox/mapbox-maps-android/blob/d5c1fdf195133f6e6719bc6beb280945c9b8f307/sdk/src/main/java/com/mapbox/maps/plugin/MapPluginRegistry.kt#L175), during our plugin initialisation phase.

> Also we don't need the view tree to find the lifecycle owner, we just need the context

I wonder what's the benefit to loop through the context recursively comparing to using in stock API for it? As [this article](https://medium.com/@takahirom/about-the-newly-added-viewtreelifecycleowner-4588e4013ca2) describes, Android has hooked the `ViewTreeLifecycleOwner` setter to `ComponentActivity` and `Fragment` and handles everything internally?

Reply:
> we don't have direct access to MapView inside the lifecycle plugin

Looking though the code, you could pass the mapView when the instance is created in `MapController#initializePlugins`, couldn't you?

> I wonder what's the benefit to loop through the context recursively comparing to using in stock API for it?

`ViewTreeLifecycleOwner` will only return the lifecycle owner of a view that is attached to a view hierarchy, won't it? That's what I see from the source code. So if we wanted to initialize the registry on creation of the plugin as suggested above, the view wouldn't yet be attached to the window and `ViewTreeLifecycleOwner.get` would return `null`. That's my understanding but it would be good to double-check.

Reply:
> I didn't notice the implementation is public (should it be?), in this case it might not make sense to introduce the constructor parameter that interferes with the registration method. 

The implementation is public so that we can initialise it in the sdk without using reflection.

> The problem is that MapboxLifecyclePluginImpl#getLifecycle will throw an uninitialized exception when called before the view is attached to the window. That said, we are not actually using the lifecycle scope created by this owner, we just rely on lifecycle callbacks to pass them down to MapView, so why don't we inject a dummy LifecycleOwner into the registry instead of MapboxLifecyclePluginImpl implementing the interface?

Sounds good, I now made the localLifecycleOwner a dummy private property instead letting MapboxLifecyclePluginImpl implementing the interface.
> 
> In the ideal situation, the MapView should be the LifecycleOwner, but since the lifecycle plugin is optional, I don't think we can do that. For v11 we could evaluate removing the lifecycle plugin and just always handling the lifecycle out-of-the-box 🙃

Yes, I'm definitely for it in v11.

Reply:
> > with this change, I can see the ON_DESTROY is not fired when the fragment is destroyed
> 
> Interesting, is the returned owner tied to the fragment or to the activity? In any case, the problem shouldn't exist if we incorporate above, but this could pose a similar problem for Nav SDK if that's the case.

@LukasPaczos I took a deeper look and the returned owner is tied to the activity it seems, and `Context.recursiveUnwrap()` returns the current context immediately. For more context, the MapFragment extends [androidx.fragment.app.Fragment](https://github.com/mapbox/mapbox-maps-android/blob/2e3e0123f6be7ff39d72d8b59c75d1760900ec0b/app/src/main/java/com/mapbox/maps/testapp/examples/fragment/MapFragment.kt#L12), it is neither a `Activity` or `ContextWrapper`, I think the `Context.recursiveUnwrap()` fall short in this use case, and probably should be revisited in the navigation SDK.

Reply:
As the above mentioned limitations, rolled back to `ViewTreeLifecycleOwner` instead of `context.toLifecyle`, and only construct the `ViewLifecycleRegistry` inside the `registerLifecycleObserver`, which is executed when the view is attached to window.